viewing_party/party.py

```diff
@@ -1,23 +1,201 @@
+    updated_watchlist = user_data
+    movie_list = []
+    movie_list.append(movie)
+
+    updated_watchlist["watchlist"] = movie_list
+
+    return updated_watchlist
+
+
+def watch_movie(user_data, movie_title):
+    user_data_dict = user_data
+
+    for movie in user_data_dict["watchlist"]:
+        if movie_title in movie.values():
+            user_data_dict["watched"].append(movie)
+            user_data_dict["watchlist"].remove(movie)
+    
+    return user_data_dict
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+
+    rating_sum = 0
+
+    watched_movies_list = user_data["watched"]
+
+    if watched_movies_list:
+        for movie in watched_movies_list:
+            rating_sum += movie["rating"]
+        return rating_sum / len(watched_movies_list)
+    else:
+        return 0
+
+
+def get_most_watched_genre(user_data):
+    
+    watched_movies_list = user_data["watched"]
+    genres_list = []
+    watched_by_genre = {}
+
+    #creates a list of genres watched, with duplicates
+    if watched_movies_list:
+        for movie in watched_movies_list:
+            genres_list.append(movie["genre"])
+    else:
+        return None
+    
+    #dictionary with genres as keys and amount of movies of that genre watched as values by counting times genre is in genres_watched list
+    for genre in genres_list:
+        watched_by_genre[genre] = genres_list.count(genre)
+    
+    top_times_watched = 0
+    top_genre = None
+    
+    # looping through dictionary to find most popular genre
+    for genre in watched_by_genre:
+        times_watched = watched_by_genre[genre]
+        if times_watched > top_times_watched:
+            top_times_watched = times_watched
+            top_genre = genre
+    
+    return top_genre
+    
+
+
+
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    
+    friends_watched_movies = set()
+    user_watched_movies = set()
+
+    # creates a set of friends' watched movies
+    for friend in user_data["friends"]:
+        for watched_list in friend["watched"]:
+            friends_watched_movies.add(watched_list['title'])
+    
+    # creates a set of user's watched movies
+    for movie in user_data["watched"]:
+        user_watched_movies.add(movie["title"])
+
+    unique_movie_titles = (user_watched_movies - friends_watched_movies)
+
+    unique_movies_data = []
+
+    #looks for unique movies in user_data and adds movie data dictionary to unique_movies_data. this could be turned into a helper function
+    for movie in unique_movie_titles:
+        for i in range(len(user_data["watched"])):
+            if movie == user_data["watched"][i]["title"]:
+                unique_movies_data.append(user_data["watched"][i])

```

Comment:
Since you have a set of titles, another way we could approach this would be to iterate over the user movies, then see whether each movie's title is in the the set. Checking for set membership is very efficient!

```python
    for movie in user_data["watched"]:
        if movie["title"] in unique_movie_titles:
            unique_movies_data.append(movie)
```